### PR TITLE
Support ORC files as data source in segment creation

### PR DESCRIPTION
**Pull request by fx19880617**
Sample commands:
build distribution:
```
mvn clean package  -DskipTests=true  -Pbin-dist
```
Run segment creation: 
```
pinot-distribution/target/apache-pinot-incubating-0.2.0-SNAPSHOT-bin/apache-pinot-incubating-0.2.0-SNAPSHOT-bin/bin/pinot-admin.sh CreateSegment -dataDir pinot-distribution/target/apache-pinot-incubating-0.2.0-SNAPSHOT-bin/apache-pinot-incubating-0.2.0-SNAPSHOT-bin/sample_data -format orc  -outDir /tmp/pinot-data/orc -tableName airlineStats -segmentName airlineStats_0 -schemaFile pinot-distribution/target/apache-pinot-incubating-0.2.0-SNAPSHOT-bin/apache-pinot-incubating-0.2.0-SNAPSHOT-bin/sample_data/airlineStats_schema.json  -overwrite
```

### CODE REVIEW

File: pinot-core/src/main/java/org/apache/pinot/core/data/recordtransformer/PinotDataType.java [hunk 1]
```diff
@@ -291,27 +304,59 @@ public Double toDouble(Object value) {
   INTEGER_ARRAY {
     @Override
     public Integer[] convert(Object value, PinotDataType sourceType) {
+      if (value instanceof List) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/data/recordtransformer/PinotDataType.java]: Please add the List support into `toIntegerArray()`, same for other types

File: pinot-core/src/main/java/org/apache/pinot/core/data/recordtransformer/PinotDataType.java [hunk 2]
```diff
@@ -291,27 +304,59 @@ public Double toDouble(Object value) {
...
+        final List list = (List) value;
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/data/recordtransformer/PinotDataType.java]: (nit) we don't use final in method

File: pinot-core/src/main/java/org/apache/pinot/core/data/recordtransformer/PinotDataType.java [hunk 3]
```diff
@@ -291,27 +304,59 @@ public Double toDouble(Object value) {
...
+        Integer[] res = new Integer[list.size()];
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/data/recordtransformer/PinotDataType.java]: cache the `list.size()`

File: pinot-core/src/main/java/org/apache/pinot/core/data/recordtransformer/PinotDataType.java [hunk 4]
```diff
@@ -263,21 +264,33 @@ public String convert(Object value, PinotDataType sourceType) {
   OBJECT {
     @Override
     public Integer toInteger(Object value) {
+      if (value instanceof Number) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/data/recordtransformer/PinotDataType.java]: Why do we need this? Number should not be categorized as OBJECT

File: pinot-orc/src/main/java/org/apache/pinot/orc/data/readers/ORCRecordReader.java [hunk 5]
```diff
@@ -164,6 +165,8 @@ private Object getBaseObject(WritableComparable w) {
       obj = ((DoubleWritable) w).get();
     } else if (BytesWritable.class.isAssignableFrom(w.getClass())) {
       obj = ((BytesWritable) w).getBytes();
+    } else if (ByteWritable.class.isAssignableFrom(w.getClass())) {
```
Comment (Jackie-Jiang) [inline pinot-orc/src/main/java/org/apache/pinot/orc/data/readers/ORCRecordReader.java]: Why `toString()`?

File: pinot-core/src/main/java/org/apache/pinot/core/data/recordtransformer/PinotDataType.java (hunk 4 above)
Comment (fx19880617) [reply pinot-core/src/main/java/org/apache/pinot/core/data/recordtransformer/PinotDataType.java]: I've seen SHORT/INTEGER/LONGs are set as OBJECT from OrcRecordReader

File: pinot-orc/src/main/java/org/apache/pinot/orc/data/readers/ORCRecordReader.java (hunk 5 above)
Comment (fx19880617) [reply pinot-orc/src/main/java/org/apache/pinot/orc/data/readers/ORCRecordReader.java]: the value is 8-bits tinyint, both `get()` and `toString()` will give same human readable numbers in a string format.

File: pinot-core/src/main/java/org/apache/pinot/core/data/recordtransformer/PinotDataType.java (hunk 4 above)
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/data/recordtransformer/PinotDataType.java]: They should fall into the corresponding PinotDataType. Please check `DataTypeTransformer.transform()`. Also I think you need to add List into that method

File: pinot-orc/src/main/java/org/apache/pinot/orc/data/readers/ORCRecordReader.java (hunk 5 above)
Comment (Jackie-Jiang) [reply pinot-orc/src/main/java/org/apache/pinot/orc/data/readers/ORCRecordReader.java]: Can you put this after the boolean type for readability?

File: pinot-core/src/main/java/org/apache/pinot/core/data/recordtransformer/PinotDataType.java [hunk 6]
```diff
@@ -355,6 +368,15 @@ public String toString(Object value) {
   }
 
   public Integer[] toIntegerArray(Object value) {
+    if (value instanceof List) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/data/recordtransformer/PinotDataType.java]: You need to use the singleValueType to do the type conversion. Please refer to the code from line 383-390, and use the same naming convention

File: pinot-orc/src/main/java/org/apache/pinot/orc/data/readers/ORCRecordReader.java (hunk 5 above)
Comment (fx19880617) [reply pinot-orc/src/main/java/org/apache/pinot/orc/data/readers/ORCRecordReader.java]: will do

File: pinot-core/src/main/java/org/apache/pinot/core/data/recordtransformer/PinotDataType.java (hunk 4 above)
Comment (fx19880617) [reply pinot-core/src/main/java/org/apache/pinot/core/data/recordtransformer/PinotDataType.java]: got it. will do

File: pinot-core/src/main/java/org/apache/pinot/core/data/recordtransformer/DataTypeTransformer.java [hunk 7]
```diff
@@ -86,9 +87,24 @@ public GenericRow transform(GenericRow record) {
...
         }
         record.putField(column, value);
+        continue;
       } else {
         PinotDataType source;
-        if (value instanceof Object[]) {
+        if (value instanceof List) {
+          // Multi-valued column
+          List values = (List) value;
+          if (!values.isEmpty()) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/data/recordtransformer/DataTypeTransformer.java]: This check should be moved to line 80

File: pinot-core/src/main/java/org/apache/pinot/core/data/recordtransformer/DataTypeTransformer.java [hunk 8]
```diff
@@ -86,9 +87,24 @@ public GenericRow transform(GenericRow record) {
...
+            source = MULTI_VALUE_TYPE_MAP.get(values.get(0).getClass());
+            if (source == null) {
+              source = PinotDataType.OBJECT_ARRAY;
+            }
+          } else {
+            source = PinotDataType.OBJECT_ARRAY;
+          }
+          value = dest.convert(value, source);
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/data/recordtransformer/DataTypeTransformer.java]: Please add some comments on why we need to convert even if source == dest

File: pinot-core/src/main/java/org/apache/pinot/core/data/recordtransformer/PinotDataType.java [hunk 9]
```diff
@@ -357,6 +358,15 @@ public String toString(Object value) {
   public Integer[] toIntegerArray(Object value) {
     if (isSingleValue()) {
       return new Integer[]{toInteger(value)};
+    } else if (value instanceof List) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/data/recordtransformer/PinotDataType.java]: Can you make it as following for readability:
```
if (isSingleValue()) {
   ...
} else {
  if (value instanceof List) {
    List valueList = (List) value;
    ...
  } else {
    Object[] valueArray = (Object[]) value;
    ...
  }
}
```

File: pinot-tools/src/main/java/org/apache/pinot/tools/admin/command/CreateSegmentCommand.java [hunk 10]
```diff
@@ -431,7 +437,25 @@ protected void getDataFilePathsHelper(FileSystem fileSystem, FileStatus[] fileSt
...
+        return fileName.endsWith(".parquet");
+      case CSV:
+        return fileName.endsWith(".csv");
+      case JSON:
+        return fileName.endsWith(".json");
+      case THRIFT:
+        return fileName.endsWith(".thrift");
+      case ORC:
+        return fileName.endsWith(".orc");
+      case OTHER:
```
Comment (Jackie-Jiang) [inline pinot-tools/src/main/java/org/apache/pinot/tools/admin/command/CreateSegmentCommand.java]: Throw exception for unsupported format?

**Pull request merged**